sample/rules_test.go

```diff
@@ -2500,7 +2500,154 @@ func TestRulesRootSpanContext(t *testing.T) {
+								Value:    "500",
+							},
+						},
+					},
+				},
+			},
+			Spans: []*types.Span{
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"http.status_code": "200",
+						},
+					},
+				},
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"http.status_code": "500",
+						},
+					},
+				},
+			},
+			ExpectedKeep: true,
+			ExpectedRate: 1,
+			ExpectedName: "no rule matched",
+		},
+		{
+			Rules: &config.RulesBasedSamplerConfig{
+				Rules: []*config.RulesBasedSamplerRule{
+					{
+						Name:       "no root span with multiple fields",
+						SampleRate: 10,
+						Conditions: []*config.RulesBasedSamplerCondition{
+							{
+								Fields:   []string{"root.foo", "foo"},
+								Operator: config.EQ,
+								Value:    100.01,
+							},
+						},
+					},
+				},
+			},
+			Spans: []*types.Span{
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"foo": 100.01,
+						},
+					},
+				},
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"foo": 99.1,
+							"bar": 100.10,
+						},
+					},
+				},
+			},
+			ExpectedKeep: false,
+			ExpectedRate: 10,
+			ExpectedName: "no root span with multiple fields",
+		},
+		{
+			Rules: &config.RulesBasedSamplerConfig{
+				Rules: []*config.RulesBasedSamplerRule{
+					{
+						Name:       "no root span, checking other spans in trace",
+						SampleRate: 10,
+						Conditions: []*config.RulesBasedSamplerCondition{
+							{
+								Fields:   []string{"root.foo", "foo"},
+								Operator: config.EQ,
+								Value:    true,
+							},
+						},
+					},
+				},
+			},
+			Spans: []*types.Span{
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"foo": true,
+						},
+					},
+				},
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"foo": true,
+						},
+					},
+				},
+			},
+			ExpectedKeep: false,
+			ExpectedRate: 10,
+			ExpectedName: "no root span, checking other spans in trace",
+		},
+		{ // practical test case where user is checking for existence of root span
+			Rules: &config.RulesBasedSamplerConfig{
+				Rules: []*config.RulesBasedSamplerRule{
+					{
+						Name:       "no root span, checking that trace id does not exist",
+						SampleRate: 10,
+						Conditions: []*config.RulesBasedSamplerCondition{
+							{
+								Fields:   []string{"root.trace_id", "trace_id"},
+								Operator: config.NotExists,
+								Value:    "abc123",
+							},
+						},
+					},
+				},
+			},
+			Spans: []*types.Span{
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"trace_id": "abc123",
+							"foo":      true,
+						},
+					},
+				},
+				{
+					Event: types.Event{
+						Data: map[string]interface{}{
+							"trace_id": "abc123",
+							"foo":      true,
+						},
+					},
+				},
+			},
+			ExpectedKeep: true,
+			ExpectedRate: 1,
+			ExpectedName: "no rule matched",
```

Comment:
Now that I see it, I don't think I understand the spirit of this test case we talked about together. 😕

Reply:
What we wanted here was to be able to tell the difference between "my root span is missing the service name" and "my root span is missing entirely".

We could do this with a two-condition test:

Condition:
    Field: root.trace_id
    Operator: exists
Condition:
    Field: root.service_name
    Operator: not-exists

That test will be true if the root span exists but the service name is missing, and false if the root span is missing entirely.

Reply:
OMG -- I forgot we added `has-root-span`!

My scenario is now kind of silly, but it doesn't hurt to have it.

Reply:
I don't think the scenario is silly! I think putting a test case with `has-root-span true` + `root.x not-exists` is the test confirmation for the advice we should give in the docs: don't use `root.x not-exists` in a rule without also including a `has-root-span true` condition or you're gonna have a bad time.

!["Gonna have a bad time" ski instructor says "That's right!"](https://media.tenor.com/GqDdOCWQBaAAAAAM/thats-right-thumper.gif)